modules/storage-api/src/test/java/org/apache/ignite/internal/storage/AbstractStorageTest.java

```diff
@@ -0,0 +1,181 @@
+
+        DataRow dataRow = dataRow("key", "value");
+
+        storage.write(dataRow);
+
+        assertArrayEquals(dataRow.value().array(), storage.read(searchRow).value().array());
+
+        storage.remove(searchRow);
+
+        assertNull(storage.read(searchRow).value());
+    }
+
+    @Test
+    public void invoke() throws Exception {
+        SearchRow searchRow = searchRow("key");
+
+        SimpleReadInvokeClosure readClosure = new SimpleReadInvokeClosure();
+
+        storage.invoke(searchRow, readClosure);
+
+        assertNull(readClosure.row().value());
+
+        DataRow dataRow = dataRow("key", "value");
+
+        storage.invoke(searchRow, new SimpleWriteInvokeClosure(dataRow));
+
+        storage.invoke(searchRow, readClosure);
+
+        Assertions.assertArrayEquals(dataRow.value().array(), readClosure.row().value().array());
+
+        storage.invoke(searchRow, new SimpleRemoveInvokeClosure());
+
+        storage.invoke(searchRow, readClosure);
+
+        assertNull(readClosure.row().value());
+    }
+
+    @Test
+    public void scanSimple() throws Exception {
+        List<DataRow> list = toList(storage.scan(key -> true));
+
+        assertEquals(emptyList(), list);
+
+        DataRow dataRow1 = dataRow("key1", "value1");
+
+        storage.write(dataRow1);
+
+        list = toList(storage.scan(key -> true));
+
+        assertThat(list, hasSize(1));
+
+        assertArrayEquals(dataRow1.value().array(), list.get(0).value().array());
+
+        DataRow dataRow2 = dataRow("key2", "value2");
+
+        storage.write(dataRow2);
+
+        list = toList(storage.scan(key -> true));
+
+        assertThat(list, hasSize(2));
+
+        // "key1" and "key2" have the same order both by hash and lexicographically.
+        assertArrayEquals(dataRow1.value().array(), list.get(0).value().array());
+        assertArrayEquals(dataRow2.value().array(), list.get(1).value().array());
+    }
+
+    @Test
+    public void scanFiltered() throws Exception {
+        DataRow dataRow1 = dataRow("key1", "value1");
+        DataRow dataRow2 = dataRow("key2", "value2");
+
+        storage.write(dataRow1);
+        storage.write(dataRow2);
+
+        List<DataRow> list = toList(storage.scan(key -> key.keyBytes()[3] == '1'));
+
+        assertThat(list, hasSize(1));
+
+        assertArrayEquals(dataRow1.value().array(), list.get(0).value().array());
+
+        list = toList(storage.scan(key -> key.keyBytes()[3] == '2'));
+
+        assertThat(list, hasSize(1));
+
+        assertArrayEquals(dataRow2.value().array(), list.get(0).value().array());
+
+        list = toList(storage.scan(key -> false));
+
+        assertTrue(list.isEmpty());
+    }
+
+    /**
+     * Converts cursor to list.
+     *
+     * @param cursor Cursor.
+     * @param <T> Type of cursor content.
+     * @return List.
+     */
+    @NotNull private <T> List<T> toList(Cursor<T> cursor) {
+        return StreamSupport.stream(cursor.spliterator(), false).collect(Collectors.toList());
```

Comment:
you actually never close the passed in cursor, so this is a resource leak

Reply:
My bad